Parser/pegen.c

```diff
@@ -410,7 +410,49 @@ get_error_line(Parser *p, Py_ssize_t lineno)
 }
 
 Py_ssize_t
-_PyPegen_byte_offset_to_character_offset(PyObject *line, Py_ssize_t col_offset)
+byte_offset_difference_for_unicode(Py_ssize_t size, PyObject *text)
+{
+    PyObject *unicode_data = PyImport_ImportModuleNoBlock("unicodedata");
```

Comment:
This is one of the reasons I was also cautious towards this approach, we either have to freeze the `unicodedata` module into the interpreter or import it, the former bloats the executable size and the latter makes it so we're introducing an import in the critical code path of execution printing.

Reply:
> introducing an import in the critical code path of execution printing.

We already import stuff like `io` at the traceback printing logic, and `unicodedata` on the parser conditionally (if we see a non-ascii identifier). We could apply the same logic to here, and check whether the line contains any non-ascii character, and use the fast path for that case. When we have unicode characters, I don't think the overall performance is going to trouble a lot considering a) C-level traceback printing logic does not generally need to be very very effective b) importing a single module (which is then cached) should not degrade the performance. But I agree that we should check on the final version and see if there are any obserable affects.

Reply:
Instead of this, I think is going to be easier, better and faster to just include this header:

https://www.cl.cam.ac.uk/~mgk25/ucs/wcwidth.c

And use that data directly, just for this purpose, as everyone else is doing (gcc, rust, clang...)

Reply:
I don't know why but the function I retrieved from that file seem to return 1 for emojis (at least for the :python:), but works fine for east asian characters. The reference it is based on suggests that

> [UTS51] emoji presentation sequences behave as though they were East Asian Wide, regardless of their assigned East_Asian_Width property value.

Which may or may not include all emojis, so I am not really sure if this is the right snippet or the table is somehow misconfigured. Any ideas?

Reply:
Checkout how rustc does it:

```
https://github.com/unicode-rs/unicode-width/blob/d0aa541144151c09268878dd781f06011498389a/src/tables.rs#L39-L48
```

Reply:
Ok, decided to go with `unicodedata` for the following 2 reasons:
- After looking at the rust implementation, I noticed that the table is not static and needs to be upgraded on each unicode release (at least those that touch on east-asian chars) and for doing so we need to create our tooling (under `Tools/unicode`) to generate a similiar table and write it off. This seems like a redundant operation considering we already have `unicodedata` that contain all the information we need.
- One major problem with this is that, I still couldn't get it to work (even after updating the table, with the command that it originally generated) on emojis. It works fine on east asian chars but not well with any sort of emojis, and I am not really sure what `unicode-width` does to deal with it so just skipping it as is. 

I've extended the current implementation include `F` (thanks to @ammaraskar) as well as `W` so I assume they should pretty much cover all cases we might need. As @pablogsal said, we probably need to test this on windows with different font combinations and check how it behaves. I also want to measure the overhead, though I don't think it will affect much.